src/python/grpcio_observability/grpc_observability/_observability.py

```diff
@@ -0,0 +1,179 @@
+}
+
+
+@dataclass
+class GcpObservabilityPythonConfig:
+    _singleton = None
+    _lock: threading.RLock = threading.RLock()
+    project_id: str = ""
+    stats_enabled: bool = False
+    tracing_enabled: bool = False
+    labels: Optional[Mapping[str, str]] = field(default_factory=dict)
+    sampling_rate: Optional[float] = 0.0
+
+    @staticmethod
+    def get():
+        with GcpObservabilityPythonConfig._lock:
+            if GcpObservabilityPythonConfig._singleton is None:
+                GcpObservabilityPythonConfig._singleton = GcpObservabilityPythonConfig(
+                )
+        return GcpObservabilityPythonConfig._singleton
+
+    def set_configuration(self,
+                          project_id: str,
+                          sampling_rate: Optional[float] = 0.0,
+                          labels: Optional[Mapping[str, str]] = None,
+                          tracing_enabled: bool = False,
+                          stats_enabled: bool = False) -> None:
+        self.project_id = project_id
+        self.stats_enabled = stats_enabled
+        self.tracing_enabled = tracing_enabled
+        self.labels = labels
+        self.sampling_rate = sampling_rate
+
+
+# pylint: disable=no-self-use
+class GCPOpenCensusObservability(grpc.ObservabilityPlugin):
+    config: GcpObservabilityPythonConfig
+    exporter: "grpc_observability.Exporter"
+
+    def __init__(self):
+        # 1. Read config.
+        self.exporter = None
+        self.config = GcpObservabilityPythonConfig.get()
+        config_valid = _cyobservability.set_gcp_observability_config(
+            self.config)
+        if not config_valid:
+            raise ValueError("Invalid configuration")
+
+        if self.config.tracing_enabled:
+            self.enable_tracing(True)
+        if self.config.stats_enabled:
+            self.enable_stats(True)
+
+    def init(self, exporter: "grpc_observability.Exporter" = None) -> None:
+        if exporter:
+            self.exporter = exporter
+        else:
+            # 2. Creating measures and register views.
+            # 3. Create and Saves Tracer and Sampler to ContextVar.
+            pass  # Actual implementation of OC exporter
+            # open_census = importlib.import_module(
+            #     "grpc_observability._open_census")
+            # self.exporter = open_census.OpenCensusExporter(
+            #     self.config.get().labels)
+
+        # 4. Start exporting thread.
+        try:
+            _cyobservability.cyobservability_init(self.exporter)
+        #TODO(xuanwn): Use specific exceptons
+        except Exception as e:  # pylint: disable=broad-except
+            _LOGGER.exception("grpc_observability init failed with: %s", e)
+
+        # 5. Init grpc.
+        # 5.1 Refister grpc_observability
+        # 5.2 set_server_call_tracer_factory
+        grpc._observability_init(self)
+
+    def exit(self) -> None:
+        # Sleep for 0.5s so all data can be flushed.
+        time.sleep(0.5)
```

Comment:
This constant is suspicious. Is there any synchronization we can do so that we don't have to sleep for an arbitrary amount of time?

Reply:
This is the time we wait for next batch in `AwaitNextBatchLocked`. And C++ is [doing similar things too](http://shortn/_huZUevSe7d). We might need change this interval to a more appropriate number later (already added TODO).

Reply:
If these two constants need to agree, then they should be pulled from a common constant so that we can't accidentally change one value but not the other.

This method isn't fool-proof. A sleep guarantees that a thread will take _at least_ that long, not exactly that long. If the thread running `AwaitNextBatchLocked` gets swapped out for some reason, then it's totally possible for the `exit` call to complete before `AwaitNextBatchLocked` finishes. The question is how big an issue that would be.

@yashykt Do you have any thoughts on this?

Reply:
I just got some clarity on this from @ejona86 . The sleep is considered a kludge. We can add some synchronization with the export thread to exit as soon as all data has been exported.